### PR TITLE
[Bug] Fix scrappy (+ some immunity move and ability) in inverse battle

### PR DESCRIPTION
**Pull request by KimJeongSun**
## What are the changes the user will see?

Scrappy, Mind's Eye abilities and the moves Foresight, Odor Sleuth, Miracle Eye no longer make moves that should deal 2x damage instead deal 1x damage in inverse battle.
These abilities and moves no longer affect type matchups in inverse battles!

## Why am I making these changes?

In Inverse Battles, these moves and abilities should not nullify type immunities (e.g., Normal -> Ghost, Psychic -> Dark), so they should have no effect. ( which are already 2x damage in inverse battle )
However, the existing code caused these moves and abilities to forcibly make the attacks deal 1x damage against these types, which is a bug.

## What are the changes from a developer perspective?

The code has been modified so that these moves and abilities do not alter type effectiveness in Inverse Mode. 
Additionally, new test cases for Inverse Battles have been added ( + including tests to confirm that Thunder Wave can affect Ground-type Pokémon ) (+ inverse battle test code refactoring! requested by @torranx )

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by Snailman11** [comment]
Closes [Bug] Scrappy prevents Super Effective attacks against Ghost types in Inverse challenge
[#3991](https://github.com/pagefaultgames/pokerogue/issues/3991)

File: src/test/battle/inverse_battle.test.ts [hunk 1]
```diff
@@ -200,4 +218,35 @@ describe("Inverse Battle", () => {
...
+    game.override
+      .ability(Abilities.SCRAPPY)
+      .enemySpecies(Species.GASTLY);
+
+    await game.challengeMode.startBattle();
+
+    const player = game.scene.getPlayerPokemon()!;
+    const enemy = game.scene.getEnemyPokemon()!;
+
+    expect(enemy.getMoveEffectiveness(player, allMoves[Moves.TACKLE])).toBe(2);
```
Comment (torranx) [inline src/test/battle/inverse_battle.test.ts:232]: you should not manually call the function but instead use spyOn on the function, for reference you can check `tera_shell.ts`

File: src/test/battle/inverse_battle.test.ts [hunk 2]
```diff
@@ -200,4 +218,35 @@ describe("Inverse Battle", () => {
...
+    await game.challengeMode.startBattle();
+
+    const player = game.scene.getPlayerPokemon()!;
+    const enemy = game.scene.getEnemyPokemon()!;
+
+    game.move.select(Moves.FORESIGHT);
+    await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
+    await game.phaseInterceptor.to(TurnEndPhase);
+
+    expect(enemy.getMoveEffectiveness(player, allMoves[Moves.TACKLE])).toBe(2);
```
Comment (torranx) [inline src/test/battle/inverse_battle.test.ts:250]: same here

**Review by torranx — COMMENTED** [review]
nit: I personally don't like how the tests are numbered, but should not block this PR

File: src/test/battle/inverse_battle.test.ts (hunk 1 above)
Comment (KimJeongSun) [reply src/test/battle/inverse_battle.test.ts:232]: Are there any advantages to using the provided code, or are there disadvantages to the existing test code method? 

When I tried it just now, the suggested method seems to be influenced by the priority of the Pokémon. Additionally, it appears impossible to check values for 0x (no effect) attacks.

Comment (KimJeongSun) [reply src/test/battle/inverse_battle.test.ts]: updated! can you review it again?

**Review by innerthunder — APPROVED** [review]

**Review by DayKev — APPROVED** [review]
The tests have some redundant code that could be removed but that can be done in a later PR.

**Pull request merged**